File: 11-container-with-most-water/11-container-with-most-water.cs
public class Solution {
public int MaxArea(int[] height) 
    {
        var h = height;
        int i = 0, j = h.Length - 1, result = 0;

        while (i < j)
        {
            result = Math.Max(result, Math.Min(h[i], h[j]) * (j - i));
            
            if (h[i] < h[j])
            {
                i++;
            }
            else 
            {
                j--;
            }
        }
        
        return result;
    }
}

File: 11-container-with-most-water/11-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int maxArea = 0;
        int start = 0, end = height.length-1;
        //two pointer strategy
        //start from extremes, and keep increasing, decreasing start end based on higher height
        while (start < end) {
            int currentArea = Math.min(height[start], height[end])*(end-start);
            maxArea = Math.max(maxArea, currentArea);
            if (height[start] < height[end]) {
                start++;
            } else {
                end--;
            }
        }
        return maxArea;
    }
}